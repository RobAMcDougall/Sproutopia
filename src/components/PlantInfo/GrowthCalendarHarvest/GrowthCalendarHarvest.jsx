import React, { useState, useEffect } from "react";
import { useParams } from "react-router-dom";
import "./../GrowthCalendar.css";

function GrowthCalendarHarvest() {
  const { id } = useParams();
  const [plant, setPlant] = useState({});

  useEffect(() => {
    async function displayGrowthData() {
      const response = await fetch(
        `https://sproutopia-backend.onrender.com/plants/${id}`
      );
      const rawData = await response.json();
      setPlant(rawData);
    }

    displayGrowthData();
  }, []);

  const getDateMonth = dateString => {
    const date = new Date(dateString);
    return date.toLocaleString("default", { month: "short" });
  };


  const isHarvestMonth = dateString => {
    const date = new Date(dateString);
    const harvestStart = new Date(plant.harvest_start);
    const harvestEnd = new Date(plant.harvest_end);

   
    return (
      (date >=
        new Date(harvestStart.getFullYear(), harvestStart.getMonth(), 1) &&
        date <= harvestEnd) ||
      (date.getMonth() === harvestStart.getMonth() &&
        date.getFullYear() === harvestStart.getFullYear())
    )
  };

  const calendarMonthsHarvest = Array.from({ length: 12 }, (_, i) => {
    const month = new Date(2024, i, 1);
    return (
      <div
        key={i}
        className={`calendar-month ${
          isHarvestMonth(month.toISOString()) ? "harvest-month" : ""
        }`}
      >
        {getDateMonth(month)}
      </div>
    );
  });
  return <div className="calendar">{calendarMonthsHarvest}</div>;
}

export default GrowthCalendarHarvest;
