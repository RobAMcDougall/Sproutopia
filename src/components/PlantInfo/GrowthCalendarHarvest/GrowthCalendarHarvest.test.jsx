import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';

import GrowthCalendarHarvest from './GrowthCalendarHarvest';

describe('GrowthCalendarHarvest', () => {
  it('renders without crashing', () => {
    render(<GrowthCalendarHarvest />);
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<GrowthCalendarHarvest />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays correct calendar months', () => {
    render(<GrowthCalendarHarvest />);
    const calendar = screen.getByTestId('calendar');
    
    
    expect(calendar.children.length).toBe(12);

   
    const months = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];
    months.forEach((month, index) => {
      expect(calendar.children[index]).toHaveTextContent(month);
    });
  });

  it('highlights harvest months correctly', () => {
    render(<GrowthCalendarHarvest />);
    const calendar = screen.getByTestId('calendar');

   
    const harvestMonthsIndices = [2, 3, 4];
    harvestMonthsIndices.forEach(index => {
      expect(calendar.children[index]).toHaveClass('harvest-month');
    });

    
    const nonHarvestMonthsIndices = [0, 1, 5, 6, 7, 8, 9, 10, 11];
    nonHarvestMonthsIndices.forEach(index => {
      expect(calendar.children[index]).not.toHaveClass('harvest-month');
    });
  });
});