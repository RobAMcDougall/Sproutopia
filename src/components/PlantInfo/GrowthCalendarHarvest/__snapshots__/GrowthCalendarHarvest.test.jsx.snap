// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GrowthCalendarHarvest > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="calendar"
  >
    <div
      class="calendar-month "
    >
      янв.
    </div>
    <div
      class="calendar-month "
    >
      февр.
    </div>
    <div
      class="calendar-month "
    >
      март
    </div>
    <div
      class="calendar-month "
    >
      апр.
    </div>
    <div
      class="calendar-month "
    >
      май
    </div>
    <div
      class="calendar-month "
    >
      июнь
    </div>
    <div
      class="calendar-month "
    >
      июль
    </div>
    <div
      class="calendar-month "
    >
      авг.
    </div>
    <div
      class="calendar-month "
    >
      сент.
    </div>
    <div
      class="calendar-month "
    >
      окт.
    </div>
    <div
      class="calendar-month "
    >
      нояб.
    </div>
    <div
      class="calendar-month "
    >
      дек.
    </div>
  </div>
</DocumentFragment>
`;
